security/src/test/java/org/kaazing/gateway/security/auth/BasicLoginModuleTest.java

```diff
@@ -135,17 +142,111 @@ public void testBasicLoginModuleIgnoresUnknownSchemes() throws Exception {
 
         context.assertIsSatisfied();
         assertNotNull(loginContext);
-        try {
-            loginContext.login();
-            final CallbackHandler nameCallbackHandler = handler.getDispatchMap().get(NameCallback.class);
-            final CallbackHandler passwordCallbackHandler = handler.getDispatchMap().get(PasswordCallback.class);
-            Assert.assertNull(nameCallbackHandler);
-            Assert.assertNull(passwordCallbackHandler);
-         } catch (LoginException e) {
-            fail("Login failed to succeed as expected: "+e.getMessage());
-        }
+
+        thrown.expect(LoginException.class);
+        loginContext.login();
+        final CallbackHandler nameCallbackHandler = handler.getDispatchMap().get(NameCallback.class);
+        final CallbackHandler passwordCallbackHandler = handler.getDispatchMap().get(PasswordCallback.class);
+        assertNull(nameCallbackHandler);
+        assertNull(passwordCallbackHandler);
+    }
+
+    @Test
+    public void testBasicLoginModuleFailsWhenNoLoginResultCallbackIsRegistered() throws Exception {
+        AuthenticationToken s = new DefaultAuthenticationToken("Basic", "Basic "+new String(Base64.encodeBase64("joe:welcome".getBytes())));
+        context.checking(new Expectations() {
+            {
+                oneOf(configuration).getAppConfigurationEntry(REALM_NAME);
+                final String loginModuleName = "org.kaazing.gateway.security.auth.BasicLoginModule";
+                final HashMap<String, Object> options = new HashMap<>();
+                final AppConfigurationEntry entry = new AppConfigurationEntry(loginModuleName,
+                        REQUIRED, options);
+                will(returnValue(new AppConfigurationEntry[]{entry}));
+            }
+        });
+
+        Subject subject = new Subject();
+        DispatchCallbackHandler handler = new DispatchCallbackHandler();
+        DefaultLoginResult loginResult = new DefaultLoginResult();
+        handler.register(AuthenticationTokenCallback.class, new AuthenticationTokenCallbackHandler(s));
+        LoginContext loginContext = new ResultAwareLoginContext("demo", subject, handler, configuration, loginResult);
+
+        context.assertIsSatisfied();
+        assertNotNull(loginContext);
+
+        thrown.expect(LoginException.class);
```

Comment:
From what I saw, the JDK will wrap most exception types into a LoginException. The handling of IoException and UnsupportedCallbackException only to throw a LoginException is consistent with what happens in the LoginModule implementations provided with the JDK and is driven by the fact that both are checked exceptions.
As a note, if there are multiple login modules in the chain and more than one fails, the LoginContext will only throw one exception, depending on the type of the login modules configured, a required login module exception taking precedence over an optional one. So, it might be helpful to log exceptions in the login module and not rely only on the single one that is thrown by the LoginContext.login() method.
You're right that some of the login module code exists that just swallows the two exceptions without providing the correct information. I'll try to update that.